### PR TITLE
Improve Graph.remove_node, add Graph.remove_subtree - WiP

### PR DESCRIPTION
**Pull request by Mikhael-Danilov**
* Graph.remove_node will remove dangling outputs
* Graph.remove_subtree method will remove a node and all its dependent elements at once

### CODE REVIEW

**Comment by ThanatosShinji** [comment]
A very interesting idea!

**Comment by ThanatosShinji** [comment]
image
I have this situation. I want to remove this Relu node in the red rectangle. According to your code, it will add Relu's next nodes to the removal list. What is supposed to deal with Conv node in the yellow rectangle? As its next node 'Add' is already removed, which makes it useless, keep it or remove it?

**Comment by ThanatosShinji** [comment]
It's an open discussion. I'm not familiar with the use case of remove_subtree. Maybe you can make the decision, as it's your need.

**Review by ThanatosShinji — APPROVED** [review]
LGTM

**Comment by Mikhael-Danilov** [comment]
> It's an open discussion. I'm not familiar with the use case of remove_subtree. Maybe you can make the decision, as it's your need.

Thank you for point. 
By my plan Conv (and any dangling) node should be removed too, so network graph will be in clean (ready to save & run) state after remove_subtree call.
I'll submit fix in a day or two.

**Comment by ThanatosShinji** [comment]
Actually, you don't need to fix it. The node that creates useless output tensors will be removed by onnx-tool. That's the Conv in the yellow rectangle. After some more model tests, this PR will be merged. Thanks for your contribution!

**Comment by ThanatosShinji** [comment]
> Actually, you don't need to fix it. The node that creates useless output tensors will be removed by onnx-tool. That's the Conv in the yellow rectangle. After some more model tests, this PR will be merged. Thanks for your contribution!

Sorry, my mistake! onnx-tool will add these useless tensors to the graph's output tensor list, which means it won't remove the dangling nodes. Maybe you need to fix it.

**Comment by Mikhael-Danilov** [comment]
After using onnx-tool some more I came up with separate Graph.remove_dangling_nodes method, which should be called after use on remove_node or remove_subtree.

**Comment by ThanatosShinji** [comment]
Great! I also think it's better to decouple two functions. As my computer is not available for about three days, I will merge this PR once it's available.

Best Regards
________________________________

**Comment by ThanatosShinji** [comment]
Is this PR ready for review?

**Comment by Mikhael-Danilov** [comment]
> Is this PR ready for review?

Sorry, not yet.

**Comment by ThanatosShinji** [comment]
Don't mind, just tell you that I'm available to review. Take your time.

Best Regards
________________________________

**Comment by Mikhael-Danilov** [comment]
I believe its ready now.

**Comment by ThanatosShinji** [comment]
@Mikhael-Danilov I tried remove_subtree with resnet50-v2-7, it seems that after the removal there is no output tensor.

File: onnx_tool/graph.py
Comment (ThanatosShinji) [inline onnx_tool/graph.py:526]: After removing it from self.output, it should append new tensors to self.output.

**Review by ThanatosShinji — CHANGES_REQUESTED** [review]
The removed graph has no output tensors.

File: onnx_tool/graph.py
Comment (Mikhael-Danilov) [reply onnx_tool/graph.py:526]: Not sure how it should be handled, on one hand this changes previous behavior, on other previous implementation leaves us with just dangling output which one are not very useful by itself.
Also there is may be case when multiple tensors stream in removed nodes, should we add new output for each?

Comment (ThanatosShinji) [reply onnx_tool/graph.py:526]: Back from vacation. Here is the simple situation: the graph has one output and the removal removes this output. The new output tensor should be the input tensor of the subtree. If the removed has no output tensor, it's not a completed onnx graph.

Comment (Mikhael-Danilov) [reply onnx_tool/graph.py:526]: Sure, yet dangling output not very useful too. Maybe we may introduce one more method `remove_dangling_outputs` to be explicitly called by user?

Comment (ThanatosShinji) [reply onnx_tool/graph.py:526]: Let's make your PR simple to merge. Could you show me your use case? Maybe a test model and some test codes. I want to make sure the generated model is a valid ONNX model. If there is no problem with your use case, let me merge your PR first. Fix the failed cases when they are issued.

Comment (ThanatosShinji) [reply onnx_tool/graph.py:526]: @Mikhael-Danilov If sharing model is not a feasible way for you, you can draw a scheme about how to remove the subtree from the graph. Anyway, the least effort to show that the code works fine for your situation.

Comment (Mikhael-Danilov) [reply onnx_tool/graph.py:526]: In my case it looks like:

`model = YOLO(modelPath)
model.export(format="onnx",imgsz=imgsz)

model = onnx_tool.Model(modelPath.replace(".pt", ".onnx"))

graph = model.graph
graph.shape_infer()

...

graph.add_dump_tensors(["/model.22/cv3.2/cv3.2.2/Conv_output_0"])
graph.remove_subtree("/model.22/Concat_2")
graph.remove_dangling_nodes()

...

model.save_model(onnxPath)`

with yolov8n as model.

So:
* I define new output
* Cut the reset
* Cleanup

Comment (ThanatosShinji) [reply onnx_tool/graph.py:526]: I see, you manually add a new output tensor. Before, we talked about the automatic method for the new output tensor. The manual way is also okay for me.

**Review by ThanatosShinji — APPROVED** [review]

**Comment by ThanatosShinji** [comment]
Thanks! @Mikhael-Danilov

**Pull request merged**
